### PR TITLE
Add an incremental resolver for minmdns usage

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
`ResolverImpl_MinMdns` has some limitations according to #18256:
  - It assumes a packet contains all data for a record. This is not true by the spec as AAAA records may be added to a packet but will not always be and must be requested separately.
  - It is stateful and unable to parse both Operational and Commission(er/able) records at the same time (but we have delegates for these at the same time)
  - It is unable to handle packets combining multiple records (e.g. thread border routers may use a single packet to reply to a browse operation for which multiple devices match)
  
#### Change overview
Add an "incremental" parser for resolving data:
   - Is initialized from an SRV record and this sets the record qname (to match TXT records) and server qname (to match A/AAAA addresses)
   - Handles receiving data packets after that, filling in information incrementally
   - Has the ability to say "IP address still required" to faciliate asking for AAAA records in the future.

This class is not yet used, however adding it and changes to it as an intermediate step as the class itself is reasonably stand alone with unit tests and can be reviewed in isolation.

#### Testing

Unit tests added for the new class.

### CODE REVIEW

**Comment by andy31415** [comment]
I still need to add some unit tests for Commissioning parsing. I believe that should not interfere with review of the code logic.

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                                                     kLocalDomain };
+
+ServiceNameType ComputeServiceNameType(SerializedQNameIterator name)
+{
+    // SRV record names look like:
+    //   <fabric>-<node>._matter._tcp.local  (operational)
+    //   <instance>._matterc._udp.local  (commissionable)
+    //   <instance>._matterd._udp.local  (commissioner)
+
+    // Get fabric/node or instance name first
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:86]: "compressed fabric id", not "fabric"

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+    ReturnErrorOnFailure(mRecordName.Set(name));
+    ReturnErrorOnFailure(mServerName.Set(srv.GetName()));
+    mCommonResolutionData.port = srv.GetPort();
+
+    {
+        // TODO: Chip code historically seems to assume that the host name is of the
+        // form "<MAC or 802.15.4 Extended Address in hex>.local" and only saves the first part.
+        //
+        // This should not be needed as server name should not be relevant once parsed.
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:160]: The hostname appears in A/AAAA records, so the comment is confusing re: "should not be needed". Also curious why you call it "serverName". "SRV == Service" not "Server"

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    switch (ComputeServiceNameType(name))
+    {
+    case ServiceNameType::kOperational:
+        mSpecificResolutionData.Set<OperationalNodeData>();
+        {
+            // Operational addresses start with peer node information
+            SerializedQNameIterator nameCopy = name;
+            if (!nameCopy.Next() || !nameCopy.IsValid())
+            {
+                mSpecificResolutionData = SpecificParseData();
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:169]: `ParseRecordSpecificData` ?

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+
+            CHIP_ERROR err =
+                ExtractIdFromInstanceName(nameCopy.Value(), &mSpecificResolutionData.Get<OperationalNodeData>().peerId);
+            if (err != CHIP_NO_ERROR)
+            {
+                mSpecificResolutionData = SpecificParseData();
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:177]: If you do this, could the request eventually await more records forever if the parsing errors and you set this empty?

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        if (data.GetName() != mRecordName.Get())
+        {
+            ChipLogDetail(Discovery, "TXT record received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+        return OnTxtRecord(data, packetRange);
+    case QType::A: {
+        if (data.GetName() != mServerName.Get())
+        {
+            ChipLogDetail(Discovery, "IP address received for a different host name.");
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:234]: Suggest just adding tracing macros here, rather than logs that may massively pollute the logs record in case of bad records

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    SerializedQNameIterator qname;
+
+    if (!ParsePtrRecord(data.GetData(), packetRange, &qname))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (qname != mRecordName.Get())
+    {
+        ChipLogDetail(Discovery, "PTR record that is not pointing to the current commission server.");
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:300]: "current commission server" is something I am not sure I understand in this context

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+        return CHIP_NO_ERROR;
+    }
+
+    // TODO: why are we not validating the string here? what is the purpose
+    // of copying and preserving the instance name here?
+    Platform::CopyString(mSpecificResolutionData.Get<CommissionNodeData>().instanceName, qname.Value());
+
+    // TODO: nothing is done with the sub name here. The instance name could be
+    //       fetched from the SRV record, so why are we processing PTR records?
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.cpp:309]: The sub name is just to make the PTR query return less result: only those matching the subtype.

File: src/lib/dnssd/IncrementalResolve.h [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    /// Return the underlying value in this object.
+    ///
+    /// Value valid as long as this object is valid and underlying Set() is
+    /// not called.
+    mdns::Minimal::SerializedQNameIterator Get() const;
+
+private:
+    // Try to have space for at least:
+    //  L1234._sub._matterc._udp.local       => 30 chars
+    //  <fabric>-<node>._mattrer._tcp.local  => 52 chars
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:56]: typo in `_matter` --> was that used to compute length?

File: src/lib/dnssd/IncrementalResolve.h [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    //  <hostname>.local (where hostname is kHostNameMaxLength == 16)
+    //
+    // This does not try to optimize out ".local" suffix which is always expected
+    // since comparisons are easier when suffix is still present.
+    static constexpr size_t kMaxStoredServerNameLength = 64;
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:61]: Not server name, it is a service name

File: src/lib/dnssd/IncrementalResolve.h [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+    uint8_t mBuffer[kMaxStoredServerNameLength] = {};
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:63]: `mNameBuffer` ?

File: src/lib/dnssd/IncrementalResolve.h [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+};
+
+/// Incrementally accumulates data from DNSSD packets. It is generally geared
+/// towards processing minmdns records.
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:67]: "towards processing minmdns records" has too much context for someone unaware to the different backends. Suggest rewording as "it is geared towards resource-constrained DNS-SD querier implementations, especially for mDNS"

File: src/lib/dnssd/IncrementalResolve.h [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    /// [data] represents the record and [packetRange] represents the range of valid bytes within
+    /// the packet for the purpose of QName parsing
+
+    CHIP_ERROR OnRecord(const mdns::Minimal::ResourceData & data, mdns::Minimal::BytesRange packetRange);
+
+    /// Return what additional data is required until the object can be extracted
+    ///
+    /// If `!GetREquiredInformation().HasAny()` the parsed information is ready
+    /// to be processed.
+    RequiredInformationFlags GetRequiredInformation() const;
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:116]: Suggest `GetMissingRequiredInformation`

File: src/lib/dnssd/IncrementalResolve.h [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+    /// Fetch the server name set by `InitializeParsing`
+    ///
+    /// Data references internal storage of this object and is valid as long
+    /// as this object is valid and InitializeParsing is not called again.
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/IncrementalResolve.h:121]: This is an ownership warning and should be made more visible

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+using namespace chip::Dnssd;
+using namespace mdns::Minimal;
+
+namespace {
+
+// Operational names must be <compressed-fabric>-<node>._matter._tcp.local
+constexpr uint8_t kTestOperationalName[] = "\0411234567898765432-ABCDEFEDCBAABCDE\07_matter\04_tcp\05local\00";
+
+// Commissionable names must be <instance>._matterc._udp.local
+constexpr uint8_t kTestCommissionableNode[] = "\020C5038835313B8B98\10_matterc\04_udp\05local\00";
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/tests/TestIncrementalResolve.cpp:42]: Please replace all octal constants with hex (`\x10` rather than `\020`).

Also suggest using string concat to show the witdths:
```
"\x10" "0C5038835313B8B98" "\x08" "_matterc" "\x04" "_udp" "\x05" "local" "\x00"
```

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+    case QType::PTR:
+        return OnPtrRecord(data, packetRange);
+    case QType::TXT:
+        if (data.GetName() != mRecordName.Get())
+        {
+            ChipLogDetail(Discovery, "TXT record received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+        return OnTxtRecord(data, packetRange);
+    case QType::A: {
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:231]: No need: enum class is not IPv4 specific and content is parsing an IPAddress. ParseARecord is the bit that has ifdef configs.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 7 above)
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:309]: We are copying it into "instance name" for specific resolution data. I am claiming this is pointless. Why are we not just ignoring the PTRs? Or if we need PTR handling it should be some specific bit, like ensuring we ask for SRV entries for the PTRs.

Currently during browse we ask for ANY and we seem to get as answers both PTR and SRV. Since we have SRV we can just go ahead and parse.

The data model here is unclear and I am complaining about it.

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+        if (!ParseTxtRecord(data.GetData(), &delegate))
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR IncrementalResolver::OnIpAddress(const Inet::IPAddress & addr)
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:336]: I don't think this is the purpose of this code. We are asking for addresses and get back addresses. If we want some sort of filtering or more logic on "append" we should add it separately. Current code does not do this and I would like to not add more functionality beyond what I intend (AAAA query and add browse support).

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestCommissionerNode));
+    NL_TEST_ASSERT(inSuite, name.Get() == AsSerializedQName(kTestCommissionableNode));
+
+    {
+        // setting to a too long value should reset it
+        uint8_t largeBuffer[256];
+        for (unsigned i = 0; i < 20; i++)
+        {
+            memcpy(largeBuffer + i * 8, "\07abcd123", 8);
+            largeBuffer[(i + 1) * 8] = 0;
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/tests/TestIncrementalResolve.cpp:133]: Just use a `chip::Encoding::BigEndian::BufferWrite` and call Put on it N times to avoid all this arithmetic

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 18]
```diff
@@ -0,0 +1,427 @@
...
+
+    NL_TEST_ASSERT(inSuite, !resolver.IsActive());
+
+    SrvRecord srvRecord;
+    PreloadSrvRecord(inSuite, srvRecord);
+
+    NL_TEST_ASSERT(inSuite, resolver.InitializeParsing(AsSerializedQName(kTestCommissionableNode), srvRecord) == CHIP_NO_ERROR);
+
+    NL_TEST_ASSERT(inSuite, resolver.IsActive());
+    NL_TEST_ASSERT(inSuite, resolver.IsActiveCommissionParse());
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/tests/TestIncrementalResolve.cpp:194]: Why is it called "CommissionParse" when it's "Commissionable" (which is a spec term). Suggest may rename to `IsActiveAndParsingCommissionable` or such

File: src/lib/dnssd/IncrementalResolve.h (hunk 8 above)
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.h:56]: Fixed typo. Correct string of:

```
'1234567898765432-ABCDEFEDCBAABCDE._matter._tcp.local'.size
=> 52
```

was used to compute size.

File: src/lib/dnssd/IncrementalResolve.h (hunk 9 above)
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.h:61]: Renamed constant.

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 19]
```diff
@@ -0,0 +1,427 @@
...
+    PreloadSrvRecord(inSuite, srvRecord);
+
+    NL_TEST_ASSERT(inSuite, resolver.InitializeParsing(AsSerializedQName(kTestCommissionableNode), srvRecord) == CHIP_NO_ERROR);
+
+    // once initialized, parsing should be ready however no IP address is available
+    NL_TEST_ASSERT(inSuite, resolver.IsActiveCommissionParse());
+    NL_TEST_ASSERT(inSuite, resolver.GetRequiredInformation().HasOnly(IncrementalResolver::RequiredInformation::kIpAddress));
+    NL_TEST_ASSERT(inSuite, resolver.GetServerName() == AsSerializedQName(kTestServerName));
+
+    // Send an irellevant IP address here
```
Comment (tcarmelveilleux) [inline src/lib/dnssd/tests/TestIncrementalResolve.cpp:318]: That's not an irrelevant IP address, that's an irrelevant **host**

File: src/lib/dnssd/IncrementalResolve.h [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+/// record so that additional DNSSD queries can be made recursively (e.g. if
+/// only a SRV/TXT records are available, ask for AAAA records).
+class IncrementalResolver
+{
+public:
+    // Elements that the incremental resolve still needs
+    enum class RequiredInformation : uint8_t
+    {
+        kSrvInitialization = 0x01, // server being initialized
+        kIpAddress         = 0x02, // IP address missing
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.h:83]: They are bit masks, however I believe here it is a subjective preference vs some objective "better/worse". 0x01 and 0x02 are obviously correct bitmasks. I find the SHL variant harder to read and more awkward.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 4 above)
Comment (tcarmelveilleux) [reply src/lib/dnssd/IncrementalResolve.cpp]: Then maybe have a `Reset()`/`ResetToInactive()` method, which avoids the mental burden of figuring-out your internal conventions.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/lib/dnssd/IncrementalResolve.cpp:343]: The issue is that this takes slots. If you have slots taken by bad addresses and don't have them left for good addresses, we have a problem.

We should have an `IPAddress::IsSupportedForMatter()` that discards bad IPv6 addresses that for sure would not interoperate

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+    NL_TEST_ASSERT(inSuite, record.Parse(srv_data, packet));
+}
+
+/// Convenience method to have a  serialized QName:
+///
+/// static const uint8_t kData[] = "datahere\00";
+///  AsSerializedQName(kData);
+///
+/// NOTE: this MUST be using the string "" format to add an extra NULL
+/// terminator that this method discards.
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp:75]: I realize this can have off-by-one but this is a test convenience helper. This is unit test code.

We do not have testing utilities for qnames. It would be easier if we would have a std::string equivalent for serializedqname and FullQName, but we do not. Adding them seemed tedious so I just took this function from other code dealing with qnames.

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp (hunk 14 above)
Comment (andy31415) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp:42]: IF I leave them in one string this gives me things like:

```
 sequence out of range [-Werror]
INFO       39 | constexpr uint8_t kTestOperationalName[] = "\x211234567898765432-ABCDEFEDCBAABCDE\x07_matter\x04_tcp\x05local\x00";
INFO          |                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO    cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
```

so then I have to split, but that also looks somewhat ugly (uses up more space).
Is this important?

Comment (andy31415) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp:42]: Spacing makes clang-format output:

```cpp
// Operational names must be <compressed-fabric>-<node>._matter._tcp.local
constexpr uint8_t kTestOperationalName[] = "\x21"
                                           "1234567898765432-ABCDEFEDCBAABCDE"
                                           "\x07"
                                           "_matter"
                                           "\x04"
                                           "_tcp"
                                           "\x05"
                                           "local"
                                           "\x00";
```

which looks terrible. So I think I lose more than I gain by using hex. Octals are hard to read beyond 7, but hopefully we can manage or one day we write qname test string classes.

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp]: Updated code mostly to not use this anymore. Testing code still needs this, but comments do not apply anymore so I removed them.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp:343]: That is fair, but still outside scope of this PR: existing code uses this logic and I am replicating it. Adding less slot usage is a separate work.

I would also argue unlikely: why would a server report such non-supported addresses?

File: src/lib/dnssd/IncrementalResolve.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.h:83]: Updated naming to be 'bit flags'.

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp (hunk 18 above)
Comment (tcarmelveilleux) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp:204]: Sure

**Review by tcarmelveilleux — APPROVED** [review]

File: src/lib/dnssd/tests/TestIncrementalResolve.cpp [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+    NL_TEST_ASSERT(inSuite, resolver.Take(nodeData) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !resolver.IsActive());
+
+    // validate data as it was passed in
+    NL_TEST_ASSERT(inSuite, nodeData.resolutionData.numIPs == 2);
+    NL_TEST_ASSERT(inSuite, nodeData.resolutionData.port == 0x1234);
+    NL_TEST_ASSERT(inSuite, !nodeData.resolutionData.supportsTcp);
+    NL_TEST_ASSERT(inSuite, nodeData.resolutionData.GetMrpRetryIntervalActive().HasValue());
+    NL_TEST_ASSERT(inSuite,
+                   nodeData.resolutionData.GetMrpRetryIntervalActive().Value() == chip::System::Clock::Milliseconds32(321));
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/tests/TestIncrementalResolve.cpp:390]: Resolving as future followup.

File: src/lib/dnssd/IncrementalResolve.cpp (hunk 4 above)
Comment (andy31415) [reply src/lib/dnssd/IncrementalResolve.cpp]: Added method and switched to RAII here.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR IncrementalResolver::OnPtrRecord(const ResourceData & data, BytesRange packetRange)
+{
+    // Here we handle subtype expectations. Data is of the form:
+    // <subtype>._sub._mattrc._udp.local or
+    // <subtype>._sub._mattrd._udp.local
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/IncrementalResolve.cpp:301]: s/mattr/matter/ in both lines, please.

File: src/lib/dnssd/IncrementalResolve.cpp [hunk 24]
```diff
@@ -0,0 +1,399 @@
...
+#include <lib/dnssd/ServiceNaming.h>
+#include <lib/dnssd/TxtFields.h>
+#include <lib/dnssd/minimal_mdns/core/RecordWriter.h>
+#include <lib/support/CHIPMemString.h>
+#include <trace/trace.h>
+
+namespace chip {
+namespace Dnssd {
+
+using namespace mdns::Minimal;
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/IncrementalResolve.cpp:28]: This file looks pretty specific to minimal mdns, but that's very non-obvious from the filename, unlike most of the other backend-specific files in this directory.
